api/users.rb

```diff
@@ -11,6 +11,65 @@
   end
 end
 
+get "#{APIPREFIX}/users/:course_id/stats" do |course_id|
+  user_data = {}
+  data = Content.collection.aggregate(
+    [
+      # Match all content in the course
+      { "$match" => { :course_id => course_id } },
+      # Keep a count of flags for each entry
+      {
+        "$set" => {
+          # Just using $ne with null will return true if the field is absent
+          # So we first fall all absent fields to null, then check if it's null,
+          # that way we match for the absence of the field or value = null
+          :is_reply => { "$ne" => [{ "$ifNull" => ["$parent_id", nil] }, nil] }
+        }
+      },
+      {
+        "$group" => {
+          # Here we're grouping items by the type (comment or thread), and the user, and whether the comment is a reply.
+          # For threads is_reply will always be false.
+          :_id => { :type => "$_type", :author_id => "$author_id", :is_reply => "$is_reply" },
+          # This will just count each group, so we get a breakdown of how many comments and threads a user has created.
+          :count => { "$sum" => 1 },
+          # These two will sum up the active and inactive reports in each category
+          # i.e. reported threads, reported comments, reported replies
+          # The way this works is (starting from inside out), we take the size of the abuse_flaggers list, we compare
+          # it to 0 using $cmp. If it's greater than zero then $cmp results in 1 otherwise 0.
+          # So we're summing up 1 for each abuse_flagger array that has entries, and 0 for the rest. This gives us a
+          # count of
+          :active_flags => { "$sum" => { "$cmp" => [{ "$size" => "$abuse_flaggers" }, 0] } },
+          :inactive_flags => { "$sum" => { "$cmp" => [{ "$size" => "$historical_abuse_flaggers" }, 0] } },
+        }
+      }
+    ])
```

Comment:
@bradenmacdonald  Would you have insight on how something like this would perform?

Reply:
@xitij2000 I don't know much about mongo performance tbh, but this query looks totally reasonable to me - a single-collection aggregation with some counts/sums. Are you worried about `{ "$size" => "$abuse_flaggers" }` or what?

You could find/make some data and test this but based on the query I wouldn't bother. If the API seems slow we can use a tool like New Relic or logging to see if it is due to this query or not, but my guess is it will be fine.

Reply:
@bradenmacdonald  My worry comes from the fact that this will operate on all the data in the course, and produce output accordingly. i.e. it will return a single unpaginated and unpaginatable json dict with an entry for each user enrolled in the course. 

In local testing with roughly about a half a dozen users and posts / comment in hundreds it's instantaneous, but I know how it'll perform with users in 5 digits and posts/comments in 6-7 digits.

Reply:
@xitij2000: My apologies, I realize I went away from this PR a while back, but I did not realize:

> it will return a single unpaginated and unpaginatable json dict with an entry for each user enrolled in the course.

That is a definite "NO". No synchronously-called API should ever do this, because the largest courses have _millions_ of students enrolled in them. The simple act of grabbing a full list of enrolled students (even without counts) can't be done as a single, unpaginated request.

Even if we could somehow generate this response in a performant way on the server side (say we asynchronously pre-rendered it into an S3 bucket and just sent back a reference to that file), we'd still be shoveling MBs of unnecessary data at the user each time. The only way I could see an unpaginated response like this working is if this feature gets automatically disabled when a course grows beyond a relatively small enrollment, but that would make it much less useful.

Reply:
@ormsbee Thanks for clarifying! I think given what you've said, I will discuss a new approach that caches the counts instead. 

To clarify a few things though: 
- This data will not be directly accessed by the user. This API will be used by edx-platform and from there the data will be sorted, paginated and then returned to the user. 
- This API is currently synchronously called since it's in testing/profiling mode, however the intention was to have this information cached in edx-platform, potentially in the DB.
- The data from edx-platform to the user would always be returned from the cache/database. 

The aim was to avoid major changes to this repo, however it seem that is the only feasible approach when dealing with such larger numbers.

Reply:
I've published a draft PR that uses a different approach: https://github.com/edx/cs_comments_service/pull/361

Instead of a huge query for all users, it builds these stats over time. The initial data is built the first time a user posts a comment, creates a thread, etc. There can be ways to backfill this as well. 

It's WIP but since this doesn't involve an aggregate query, we can easily do sorting and pagination with that (not implemented yet).